src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java

```diff
@@ -0,0 +1,328 @@
+import org.broadinstitute.hellbender.cmdline.CommandLineProgramProperties;
+import org.broadinstitute.hellbender.cmdline.StandardArgumentDefinitions;
+import org.broadinstitute.hellbender.cmdline.programgroups.SparkProgramGroup;
+import org.broadinstitute.hellbender.engine.spark.GATKSparkTool;
+import org.broadinstitute.hellbender.engine.spark.datasources.ReadsSparkSink;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.gcs.BucketUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadsWriteFormat;
+import org.broadinstitute.hellbender.utils.read.SAMRecordToGATKReadAdapter;
+import org.seqdoop.hadoop_bam.BAMInputFormat;
+import scala.Tuple2;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.*;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.function.Function;
+
+import static org.broadinstitute.hellbender.utils.Utils.calcMD5;
+
+@CommandLineProgramProperties(summary = "Runs BWA",
+        oneLineSummary = "BWA on Spark",
+        programGroup = SparkProgramGroup.class)
+public final class BwaSpark extends GATKSparkTool {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Logger log = LogManager.getLogger(BwaSpark.class);
+
+    private static ThreadLocal<Stopwatch> stopwatchThreadLocal = new ThreadLocal<Stopwatch>() {
+        @Override
+        protected Stopwatch initialValue() {
+            return Stopwatch.createUnstarted();
+        }
+    };
+
+    private static Set<Stopwatch> stopwatches = new HashSet<>();
+
+    @Argument(doc = "the reference", shortName = "ref",
+            fullName = "ref", optional = false)
+    private String ref;
+
+    @Argument(doc = "the output bam", shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, optional = false)
+    private String output;
+
+    @Argument(doc = "the number of base pairs to send in a batch to BWA", shortName = "K",
+            fullName = "fixedChunkSize", optional = true)
+    private int fixedChunkSize = 100000;
+
+    @Argument(doc = "the number of threads", shortName = "t",
+            fullName = "threads", optional = true)
+    private int numThreads = 1;
+
+    @Override
+    public boolean requiresReads() {
+        return true;
+    }
+
+    @Override
+    protected void runTool(final JavaSparkContext ctx) {
+        Stopwatch toolStopwatch = Stopwatch.createStarted();
+        ctx.hadoopConfiguration().setBoolean(BAMInputFormat.KEEP_PAIRED_READS_TOGETHER_PROPERTY, true);
+        JavaRDD<GATKRead> unalignedReads = getReads();
+        JavaRDD<List<GATKRead>> unalignedPairs = unalignedReads.mapPartitions(iter -> () -> pairwise(iter));
```

Comment:
should this call to `pairwise` just be inlined? It may make it simpler to read I think